src/algorithms/ctmrg_all_sides.jl

```diff
@@ -0,0 +1,177 @@
+    ϵ = 0.0
+    drc_combinations = collect(Iterators.product(axes(env.corners)...))
+    @fwdthreads for (dir, r, c) in drc_combinations
+        # Row-column index of next enlarged corner
+        next_rc = if dir == 1
+            (r, _next(c, size(env.corners, 3)))
+        elseif dir == 2
+            (_next(r, size(env.corners, 2)), c)
+        elseif dir == 3
+            (r, _prev(c, size(env.corners, 3)))
+        elseif dir == 4
+            (_prev(r, size(env.corners, 2)), c)
+        end
+
+        # SVD half-infinite environment
+        trscheme = if T <: FixedSpaceTruncation
+            truncspace(space(env.edges[dir, r, c], 1))
+        else
+            alg.trscheme
+        end
+        svd_alg = if A <: SVDAdjoint{<:FixedSVD}
+            idx = (dir, r, c)
+            fwd_alg = alg.svd_alg.fwd_alg
+            fix_svd = FixedSVD(fwd_alg.U[idx...], fwd_alg.S[idx...], fwd_alg.V[idx...])
+            SVDAdjoint(; fwd_alg=fix_svd, rrule_alg=alg.svd_alg.rrule_alg)
+        else
+            alg.svd_alg
+        end
+        @autoopt @tensor QQ[χ_in D_inabove D_inbelow; χ_out D_outabove D_outbelow] :=
+            Q[dir, r, c][χ_in D_inabove D_inbelow; χ D1 D2] *
+            Q[_next(dir, 4), next_rc...][χ D1 D2; χ_out D_outabove D_outbelow]
+        U_local, S_local, V_local, ϵ_local = PEPSKit.tsvd!(QQ, svd_alg; trunc=trscheme)
+        U[dir, r, c] = U_local
+        S[dir, r, c] = S_local
+        V[dir, r, c] = V_local
+        ϵ = max(ϵ, ϵ_local / norm(S_local))
+
+        # Compute SVD truncation error and check for degenerate singular values
+        ignore_derivatives() do
+            if alg.verbosity > 0 && is_degenerate_spectrum(S_local)
+                svals = TensorKit.SectorDict(c => diag(b) for (c, b) in blocks(S_local))
+                @warn("degenerate singular values detected: ", svals)
+            end
+        end
+
+        # Compute projectors
+        Pl, Pr = build_projectors(
+            U_local, S_local, V_local, Q[dir, r, c], Q[_next(dir, 4), next_rc...]
+        )
+        P_left[dir, r, c] = Pl
+        P_right[dir, r, c] = Pr
+    end
+
+    return copy(P_left), copy(P_right), (; ϵ, U=copy(U), S=copy(S), V=copy(V))
+end
+
+# Apply projectors to renormalize corners and edges
+function renormalize_corners_edges(state, env::CTMRGEnv, Q, P_left, P_right)
+    corners::typeof(env.corners) = copy(env.corners)
+    edges::typeof(env.edges) = copy(env.edges)
+    rc_combinations = collect(Iterators.product(axes(state)...))
+    @fwdthreads for (r, c) in rc_combinations
+        rprev = _prev(r, size(state, 1))
+        rnext = _next(r, size(state, 1))
+        cprev = _prev(c, size(state, 2))
+        cnext = _next(c, size(state, 2))
+        @diffset @autoopt @tensor corners[NORTHWEST, r, c][χ_S; χ_E] :=
+            P_right[WEST, rnext, c][χ_S; χ1 D1 D2] *
+            Q[NORTHWEST, r, c][χ1 D1 D2; χ2 D3 D4] *
+            P_left[NORTH, r, c][χ2 D3 D4; χ_E]
```

Comment:
Wow this actually took me way too long since `@diffset` behaved really weirdly. Apparently
```julia
C_NW = _contract_new_corner(
            P_right[WEST, rnext, c], Q[NORTHWEST, r, c], P_left[NORTH, r, c]
)
@diffset corners[NORTHWEST, r, c] = C_NW
```
produces correct results but assigning directly to the destination
```julia
@diffset corners[NORTHWEST, r, c] = _contract_new_corner(
            P_right[WEST, rnext, c], Q[NORTHWEST, r, c], P_left[NORTH, r, c]
)
```
somehow produces wrong results but also doesn't error. What a headache... At some point we should maybe really try to scrap `@diffset`

Reply:
Upon further inspection, I found that this has to do with the parallelization together with `@diffset`. I'm not sure what is happening but somehow `@diffset` together with `Threads.@threads` must produces some sort of race condition. (Without `@diffset` there definitely shouldn't be a race condition.)